### DISCUSSION TITLE
Word Break II

### DISCUSSION THREAD

**Discussion by ghost** [community/community#138507] [Programming Help] [answered]
### Body

Given a string s and a dictionary of words wordDict, you need to return all possible sentences that can be constructed from the words in the dictionary where each sentence uses every word in wordDict exactly once. Each sentence should be formed by concatenating words from wordDict such that the resulting sentence is a substring of s.

**_Requirements:_**

1. **Function Signature:**

`def word_break(s: str, wordDict: List[str]) -> List[str]:`

2. **Example:**

`word_break("catsanddog", ["cat", "cats", "and", "sand", "dog"])`
`# Output: ["cat sand dog", "cats and dog"]`

`word_break("pineapplepenapple", ["apple", "pen", "applepen", "pine", "pineapple"])`
`# Output: ["pine apple pen apple", "pine applepen apple", "pineapple pen apple"]`

3. **Constraints:**

- The input string s consists of lowercase English letters.
- The wordDict contains unique words and may have duplicates. The length of wordDict is at most 1000.
- The length of the input string s will be at most 1000.

**_Approach:_**

Consider using dynamic programming combined with backtracking:

1. **Dynamic Programming (DP) Array:** Create a DP array dp where dp[i] holds a list of sentences that can be formed from the substring s[0:i].

2. **Backtracking:** Use backtracking to construct sentences from the words in wordDict that match the current substring.
 
3. **Recursive Function:** Implement a recursive function that builds the sentences from substrings of s using words in wordDict and stores the results in the DP array.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by MatheusFigueiredo7** [comment, marked as answer]
```
from typing import List

def word_break(s: str, wordDict: List[str]) -> List[str]:
    word_set = set(wordDict)
    memo = {}
    
    def backtrack(start):
        if start in memo:
            return memo[start]
        
        res = []
        if start == len(s):
            res.append("")
        
        for end in range(start + 1, len(s) + 1):
            word = s[start:end]
            if word in word_set:
                for sentence in backtrack(end):
                    if sentence:
                        res.append(word + " " + sentence)
                    else:
                        res.append(word)
        
        memo[start] = res
        return res
    
    return backtrack(0)

```
